KEP-3716: Admission Webhook Match Conditions (#3717)
kubernetes/enhancements: keps/sig-api-machinery/3716-webhook-predicates/README.md

	// 'oldObject' - The existing object. The value is null for CREATE requests.
	// 'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
	//
	// Required.
	Expression string `json:"expression"`
}
```

The match condition expression is evaluated by the same libraries as those used for CEL
ValidatingAdmissionPolicy. The only difference in expressions is the availability of the `params`
variable. Expressions requiring access to additional information outside the AdmissionRequest must
be performed in the webhook, and are out of scope for this proposal.

### Risks and Mitigations

#### Security

**Risk: Attacker adds or changes a match condition to weaken an admission policy.**

This is does not represent a new threat, as doing so would require update access to the admission
registration object, and with that permission an attacker could already disable the policy through
manipulating match rules, namespace selector, or object selector (or reroute the webhook entirely).

**Risk: Logic error in match condition expression.**

Currently the match conditions must be encoded in the webhook backend itself. Moving the logic

> into a CEL expression does not materially increase the risk of a logic bug.

Reviewer: Can you unit-test CEL expressions against input yamls? I'd argue that ease of testing does materially decrease the risk of a logic bug.

Author: Good point. I think the cell ecosystem is currently missing a lot of tools, so this is something that could be made a lot easier. I'll add a comment about it.

is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `AdmissionWebhookMatchConditions`
  - Components depending on the feature gate: `kube-apiserver`

###### Does enabling the feature change any default behavior?

No. If the feature is enabled, but the `matchConditions` field is unset, the default behavior
remains unchanged.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Disabling the feature gate will ignore any `matchConditions` set, and return to the default
behavior. This could increase the traffic to the webhook, and potentially increase the error rate if

> the webhook is down or rejects those requests.

Reviewer: Interestingly, it could be possible to set up `matchConditions` in such a way that when the feature is disabled, makes the cluster more unstable. A webhook (from above) like this:

```yaml
rules:
  - operations: ['CREATE']
    apiGroups: '' # core
    apiVersions: '*'
    resources: 'pods'
matchConditions:
  # Only include pods with an NFS volume.
  - expression: 'object.spec.volumes.exists(v, v.has(nfs))'
```

will only trigger when a pod has an NFS volume. But let's imagine disabling this feature for a sec. Then this webhook gets executed for all pods. If there is some error in the logic of the webhook, i.e. it presumes the existence of an NFS volume, the webhook could just end up exploding along that path and end up rejecting all pod requests along with it.

I'm not sure that you can do anything about this, but it's worth pointing out I think.

Author: Yeah, that's what I was implying with

> This could increase the traffic to the webhook, and potentially increase the error rate if
> the webhook is down or rejects those requests.

I'll clarify it a bit.

and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

> - [ ] Metrics

Reviewer: A counter which showed us a webhook and the number of requests that were exempted could probably be useful.

Author: Where would you put that? I was under the impression that we couldn't add webhook name as a label for metric cardinality, and I'm not sure how useful a global counter would be. WDYT?

Reviewer: We already include webhook names in our webhook metrics (https://github.com/kubernetes/kubernetes/blob/9d0d2e8ece9bdd0cd8c23be2f36eee5473afc648/staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go#L172-L188). While theoretically, webhook names are unbounded, in practice the apiserver can only have so many webhooks in a chain before it just stops working, so there tends to be a limit in practice. 

CEL probably increases this limit drastically though. But we'll only really start bumping into cardinality concerns if we have more than like a thousand webhooks installed simultaneously.

Author: Oh, interesting. I didn't realize that. In that case, I agree that this makes sense.

Author: Done. Added the metric spec under `Monitoring Requirements`. Advice on the metric name is welcome...

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes adding "match conditions" to admission webhooks, as an extension to the
existing `rules` to define the scope of a webhook. A `matchCondition` is a
[CEL](https://github.com/google/cel-spec) expression that must evaluate to true for the admission
request to be sent to the webhook. If a `matchCondition` evaluates to false, the webhook is skipped for
that request (implicitly allowed).

## Motivation

**Reliability:** Admission webhooks continue to be an operational sore spot for many Kubernetes
users. Webhooks that target cluster critical resources put the admission controller backing the
webhook in the critical path of cluster stability. Even if tools like namespace scoping are used to
avoid circular-dependencies and exclude critical system resources, a webhook outage can still have a
major impact on cluster availability. This proposal aims to mitigate (but not eliminate) these
issues by allowing webhooks to be more narrowly scoped and targeted.

The same benefits also apply to bootstrapping a new cluster, or in a disaster-recovery scenario

> where objects in a cluster need to be recreated from source code. Being able to define and deploy a

Reviewer 2: I don't understand how this proposal gets you the ability to deploy a configuration which routes requests to pods that aren't running yet. I don't think you meant to say that but that's how this reads to me.

`ValidatingAdmissionPolicy` is an exciting new feature which we hope will greatly reduce the need
for admission webhooks, but it is intentionally not attempting to cover every possible use case.
This proposal aims to improve the situation for those webhooks that cannot be migrated.

### User Stories

#### Exclude resources from a wildcard rule

> I want to enforce metadata policy through an admission webhook without adding latency & risk to
> high QPS system requests.

Currently, if a webhook uses wildcard match rules, there is no way to filter out a subset of
resources or requests from matching the wildcard. If the webhook instead enumerates every resource
that should match, it must be kept up-to-date with every CRD that's added.

With CEL match conditions, the webhook could specify wildcard match rules, and add match conditions
to filter out the desired resources:

```yaml
rules:
  # Match CREATE & UPDATE on all resources:
  - operations:
    - CREATE
    - UPDATE
    apiGroups: '*'
    apiVersions: '*'
    resources: '*'
matchConditions:
  # Exclude leases from the webhook

>   - expression: '!(request.resource.group == "coordination.k8s.io" && resource.resource == "leases")'

Reviewer 2: I suggest you make these {name, expression} pairs. That way it would be easy to have a controller own and maintain one or more named expressions. The name helps explain what the expression is for, and it make SSA work well. The name could also be reported/logged in errors etc.

> requests.

System _resources_ can currently be exempted through a namespace or label selector, but requests by
system components against non-system resources cannot be. For example, update pod status requests by
Kubelets cannot be excluded from user webhooks intercepting all pod requests.

With `matchConditions`, a managed cluster could append system-exclusion rules to each webhook. For example:

```yaml
matchConditions:
  # Exclude node requests from the webhook
  - expression: '!("system:nodes" in request.userInfo.groups)'
```

Since the expression will be evaluated using a common Kubernetes CEL library, these expressions
should also get automatic access to the secondary authorization check mechanism described in
[KEP-3488: CEL for Admission Control](/keps/sig-api-machinery/3488-cel-admission-control#secondary-authz).
In practice, this means that RBAC bindings can be used to opt-out privileged users from security policy:

_Note: The secondary authz mechanism has not yet been designed, and the example syntax here is just
to illustrate how it might be used._

```yaml
matchConditions:
  # Exclude users with the 'breakglass' permission on the 'security-policy' webhook.

>   - expression: '!authorized(request.userInfo, "breakglass", "validatingwebhookconfigurations.admissionregistration.k8s.io/security-policy")'

Reviewer 2: This seems like something either a) the people with breakglass permission or b) the cluster admin might want to add to many webhook configurations. But: there's no facility to make it easy or possible for either of those groups to require this of webhook configurations. Is there some further plan?

Author: I would definitely like to explore match conditions that are applied across webhooks. Another use case to add to the 2 you listed is if a managed cluster provider wants to exclude managed resources from webhooks. @soorena776 is also looking into this.

I decided to make this out-of-scope for this KEP, since I think it expands the scope quite a bit, but we should definitely discuss it in the future.

I added a `Future work` section.

ValidatingAdmissionPolicy. The only difference in expressions is the availability of the `params`
variable. Expressions requiring access to additional information outside the AdmissionRequest must
be performed in the webhook, and are out of scope for this proposal.

### Risks and Mitigations

#### Security

**Risk: Attacker adds or changes a match condition to weaken an admission policy.**

This is does not represent a new threat, as doing so would require update access to the admission
registration object, and with that permission an attacker could already disable the policy through
manipulating match rules, namespace selector, or object selector (or reroute the webhook entirely).

**Risk: Logic error in match condition expression.**

Currently the match conditions must be encoded in the webhook backend itself. Moving the logic into
a CEL expression adds a potential failure point. This can be mitigated by testing, but the CEL
ecosystem currently lacks some of the tools that would make this easier.

Of particular significance are match conditions tied to non-functional properties of an object, such

> as using labels to decide whether to opt an object out of a policy. Without additional admition

Reviewer 2: ```suggestion
as using labels to decide whether to opt an object out of a policy. Without additional admission
```

Author: :flushed:

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Complete features A, B, C
- Additional tests are in Testgrid and linked in KEP

#### GA

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

> ### Upgrade / Downgrade Strategy

Reviewer 3: Downgrade has interesting enforcement properties on admission webhooks that previously had limited scope.  It's worth a mention.

Author: Done.

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

> ### Version Skew Strategy

Reviewer 3: skew has similar problems with over-enforcement as above, also worth a mention.

Author: Done.

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

> ###### What steps should be taken if SLOs are not being met to determine the problem?

Reviewer 3: Link to your debugability section here please.

Author: Done.

> # KEP-3716: Webhook Predicates

Reviewer 4: ```suggestion
# KEP-3716: Admission Webhook Predicates
```

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes adding "predicate" expressions to admission webhooks, as an extension to the
existing `rules` to define the scope of a webhook. A `predicate` is a
[CEL](https://github.com/google/cel-spec) expression that must evaluate to true for the admission
request to be sent to the webhook. If a `predicate` evaluates to false, the webhook is skipped for
that request (implicitly allowed).

<<[UNRESOLVED naming ]>>
I'm not satisfied with the name `predicates` for this feature. Other ideas considered:
- filters, prefilters

> - match{Predicates, Filters, Criteria, Expressions, Constraints, ...}

Reviewer 4: Conditions?

Author: Yeah, over on https://github.com/kubernetes/enhancements/pull/3697#discussion_r1063011479 we landed on Preconditions as a good option. I might change this PR to use that instead.

Reviewer 5: +1 to preconditions

Reviewer 6: `matchExpressions` was my 1st choice, but I can get behind `preconditions` if that resonates with people

Author: FYI, we're also considering `matchConditions` on https://github.com/kubernetes/enhancements/pull/3697#discussion_r1063011479

Author: I pushed an update going with `matchConditions`, but we can still decide to change it.

<<[/UNRESOLVED]>>

## Motivation

**Reliability:** Admission webhooks continue to be an operational sore spot for many Kubernetes
users. Webhooks that target cluster critical resources put the admission controller backing the
webhook in the critical path of cluster stability. Even if tools like namespace scoping are used to
avoid circular-dependencies and exclude critical system resources, a webhook outage can still have a
major impact on cluster availability. This proposal aims to mitigate (but not eliminate) these
issues by allowing webhooks to be more narrowly scoped and targeted.

> 

Reviewer 4: ```suggestion

The same benefits also apply to bootstrapping a new cluster, or in a disaster-recovery scenario where
objects in a cluster need to be recreated from source code. Being able to define and deploy a webhook before
any workload Pods are defined means that admins don't need to bypass security controls to fix things, and
helps to ensure there's never a security gap when restoring service.

```
?

**Performance:** Admission webhooks sit in the critical request path for write-requests. Validating
webhooks can be run in parallel, but Mutating webhooks must be run in serial (up to 2 times!). This
makes webhooks extremely latency sensitive, and even a webhook that doesn't do any work still needs
to pay the network round-trip cost.

**Supportability:** For hosted or managed Kubernetes distributions, webhooks can be a problem when
they interfere with requests by managed components. The existing criteria for filtering out requests
are insufficient for many use cases, and aren't easily appended with provider rules.

_What about [CEL for Admission Control](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control)?_
`ValidatingAdmissionPolicy` is an exciting new feature which we hope will greatly reduce the need
for admission webhooks, but it is intentionally not attempting to cover every possible use case.
This proposal aims to improve the situation for those webhooks that cannot be migrated.

### User Stories

> 

Reviewer 4: The webhook implementation of Pod Security Admission is a good example here I think: you might prefer to be able to run a healthy control plane in `kube-system` even if the otherwise-mandatory PSA webhook wasn't healthy.

Author: You can do that today with a namespace selector though. In fact, since PSA operates on namespace labels, you could exclude any namespace labeled as `privileged`.

I chose the use cases here specifically because they're things that don't currently have a good solution.

that the request is within scope, and return early if it's not. This adds latency and an additional
failure point to irrelevant requests. This example requires an external integration, and thus is not
a candidate for migration to CEL `ValidatingAdmissionPolicy`.

With predicates, the predicate expressions can check whether the request object is in-scope for the
webhook:

```yaml
rules:
  - operations: ['CREATE']
    apiGroups: '' # core
    apiVersions: '*'
    resources: 'pods'
predicates:
  # Only include pods with an NFS volume.
  - expression: 'request.object.spec.volumes.exists(v, v.has(nfs))'
```

### Goals

1. Provide a filtering mechanism for excluding requests from an admission webhook
2. Maintain consistency with `ValidatingAdmissionPolicy`

### Non-Goals

* Provide a mechanism to exclude requests from all webhooks.

## Proposal

### API

Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
with a new `Predicates` field:

```go

type ValidatingWebhook struct {
  // ...

  // Predicates is a list of conditions on the AdmissionRequest ('request') that must be met for a
  // request to be sent to this webhook.
  // +optional

>   Predicates []Predicate `json:"predicates,omitempty"`

Reviewer 6: `ruleExpressions`?

	// request.object. No other metadata properties are accessible.
	//
	// Only property names of the form `[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible.
	// Accessible property names are escaped according to the following rules when accessed in the expression:
	// - '__' escapes to '__underscores__'
	// - '.' escapes to '__dot__'
	// - '-' escapes to '__dash__'
	// - '/' escapes to '__slash__'
	// - Property names that exactly match a CEL RESERVED keyword escape to '__{keyword}__'. The keywords are:
	//	  "true", "false", "null", "in", "as", "break", "const", "continue", "else", "for", "function", "if",
	//	  "import", "let", "loop", "package", "namespace", "return".
	// Examples:
	//   - Expression accessing a property named "namespace": {"Expression": "object.__namespace__ > 0"}
	//   - Expression accessing a property named "x-p
<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->
```

> The predicate expression has access to the contents of the `AdmissionRequest` object (exposed as the

Reviewer 6: For consistency with ValidatingAdmissionPolicies, make 'object' a top level variable? This will also make it easier to consolidate code between policies and webhooks.

Reviewer 5: +1 -- I would also make a case for `oldObject` to be a top level variable for mutation cases.

`request` variable), but is not given any additional information. Expressions requiring access to

> additional information (such as a paramater object) must be performed in the webhook, and are out of

Reviewer 6: I agree with this. In particular, I don't think we need parameter objects here.  I don't see this as an inconsistency between webhooks and policies since parameters do not exist on all policies..

scope for this proposal.

<<[UNRESOLVED api structure ]>>

> Do we need the predicate struct? Or should we just inline it instead? That is:

Reviewer 6: Using a struct is slightly more future proof. If we end up having to add adjacent fields we can.  Concretely, if we ever decided we wanted to offer a language alternative to CEL (or replace CEL) we could need a `language` field. This would give us a place to put it.

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

> 

Reviewer 4: Instead of a `predicates:` field, how about an `exclusions:` field?

```yaml
exclusions:
  # Always admit Leases
  - expression: >-
     (request.resource.group == "coordination.k8s.io" && resource.resource == "leases")
```

I'm mentioning this not because I prefer this choice but because adding it lets us outline the rationale (which I think is that Kubernetes APIs prefer to declaratively define what's valid rather than what isn't).

Author: Exclusions does capture the intent of the KEP, but it also inverts the meaning of the CEL expressions in a way that makes them more confusing, IMO.

Reviewer 4: Yes. We should _not_ use `exclusions`; we should (I propose) list it as an alternative that we rejected, and explain why.

### Resource Exclusions

> https://github.com/kubernetes/enhancements/pull/3694 Proposes an alternative approach using a more

Reviewer 4: ```suggestion
[KEP-3693](https://github.com/kubernetes/enhancements/issues/3693) proposes an alternative approach using a more
```

## Summary

This KEP proposes adding "match conditions" to admission webhooks, as an extension to the
existing `rules` to define the scope of a webhook. A `matchCondition` is a
[CEL](https://github.com/google/cel-spec) expression that must evaluate to true for the admission
request to be sent to the webhook. If a `matchCondition` evaluates to false, the webhook is skipped for
that request (implicitly allowed).

## Motivation

**Reliability:** Admission webhooks continue to be an operational sore spot for many Kubernetes
users. Webhooks that target cluster critical resources put the admission controller backing the
webhook in the critical path of cluster stability. Even if tools like namespace scoping are used to
avoid circular-dependencies and exclude critical system resources, a webhook outage can still have a
major impact on cluster availability. This proposal aims to mitigate (but not eliminate) these
issues by allowing webhooks to be more narrowly scoped and targeted.

The same benefits also apply to bootstrapping a new cluster, or in a disaster-recovery scenario
where objects in a cluster need to be recreated from source code. Being able to define and deploy a

> webhook before any workload Pods are defined means that admins don't need to bypass security

Reviewer 3: With careful use of namespace, I would expect this to be possible today.  It doesn't invalidate the other purposes of this feature, but I want to be clear on what is unique value here, since the mechanisms very closely parallel existing mechanisms.

Author: @sftim proposed this change, WDYT?

I mostly agree with @deads2k that namespace selectors might be better suited for this purpose, although I could imagine subject rules being useful in this situation too.

`ValidatingAdmissionPolicy` is an exciting new feature which we hope will greatly reduce the need
for admission webhooks, but it is intentionally not attempting to cover every possible use case.
This proposal aims to improve the situation for those webhooks that cannot be migrated.

### User Stories

#### Exclude resources from a wildcard rule

> I want to enforce metadata policy through an admission webhook without adding latency & risk to
> high QPS system requests.

Currently, if a webhook uses wildcard match rules, there is no way to filter out a subset of
resources or requests from matching the wildcard. If the webhook instead enumerates every resource
that should match, it must be kept up-to-date with every CRD that's added.

With CEL match conditions, the webhook could specify wildcard match rules, and add match conditions
to filter out the desired resources:

```yaml
rules:
  # Match CREATE & UPDATE on all resources:
  - operations:
    - CREATE
    - UPDATE
    apiGroups: '*'
    apiVersions: '*'
    resources: '*'
matchConditions:
  # Exclude leases from the webhook
  - expression: '!(request.resource.group == "coordination.k8s.io" && resource.resource == "leases")'
```

> #### Exempt system users from security policy

Reviewer 3: This is a potentially useful capability for something like nodes, but using it like this requires a binding (potentially at the level of namespace) to be made inside of the admission configuration.  In validating cel admission we identified the need to run an authorization check to satisfy a user level permissions check while allowing existing binding and extensibility techniques to be utilized.  I think similar reasons exist here.

Can you add a use-case for, "user has permission to delete resources" as a reason to skip a check.  This would be similar to checks we make inside the GC admission plugin before performing an evaluation.

Reviewer 3: > I'm not completely against the idea, but I'd like to see that pattern well established in ValidatingAdmissionPolicy before we applied it here.

We can talk in high bandwidth at the sig meeting, but if the CEL expression evaluation consistently with validating admission policy, this should "just happen". 

If it cannot be included, I have reservations about introducing an additional permissions binding mechanism that cannot re-use our existing patterns and extensions.  I'll also have questions about how we intended to keep our CEL expression language consistent if additions in validating admission policy do not also appear in webhook match conditions.

Author: OK, I added a reference for this, and a sample user story.

Reviewer 3: That helps, thanks.

> As a managed cluster provider, I want to prevent user webhooks from intercepting critical system
> requests.

System _resources_ can currently be exempted through a namespace or label selector, but requests by
system components against non-system resources cannot be. For example, update pod status requests by
Kubelets cannot be excluded from user webhooks intercepting all pod requests.

With `matchConditions`, a managed cluster could append system-exclusion rules to each webhook. For example:

```yaml
matchConditions:
  # Exclude node requests from the webhook
  - expression: '!("system:nodes" in request.userInfo.groups)'
```

> #### Scope an NFS access management webhook to Pods mounting NFS volumes

Reviewer 3: I found this use-case and the permissions based use-cases to be the most compelling.

> pods with NFS volumes need to be checked.

Currently, there is no way to achieve this. Many webhook implementations today start by checking
that the request is within scope, and return early if it's not. This adds latency and an additional
failure point to irrelevant requests. This example requires an external integration, and thus is not
a candidate for migration to CEL `ValidatingAdmissionPolicy`.

With match conditions, the expressions can check whether the request object is in-scope for the
webhook:

```yaml
rules:
  - operations: ['CREATE']
    apiGroups: '' # core
    apiVersions: '*'
    resources: 'pods'
matchConditions:
  # Only include pods with an NFS volume.
  - expression: 'request.object.spec.volumes.exists(v, v.has(nfs))'
```

### Goals

1. Provide a filtering mechanism for excluding requests from an admission webhook
2. Maintain consistency with `ValidatingAdmissionPolicy`

### Non-Goals

* Provide a mechanism to exclude requests from all webhooks.

## Proposal

### API

Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
with a new `MatchConditions` field:

```go

type ValidatingWebhook struct {
  // ...

>   // MatchConditions is a list of conditions on the AdmissionRequest ('request') that must be met for a

Reviewer 3: Please add a clarifying sentence that these MatchConditions are AND'd.  Reading top to bottom I think that's the intent here.

Author: Done.

  // request to be sent to this webhook.
  // +optional
  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
}

type MutatingWebhook struct {
  // ...
  MatchConditions []MatchCondition `json:"matchConditions,omitempty"`
}

// MatchCondition represents a condition which must by fulfilled for a request to be sent to a webhook.
type MatchCondition struct {
	// Expression represents the expression which will be evaluated by CEL.
	// ref: https://github.com/google/cel-spec
	// CEL expressions have access to the contents of the AdmissionRequest, organized into CEL variables:
	//
	//'object' - The object from the incoming request. The value is null for DELETE requests.
	//'oldObject' - The existing object. The value is null for CREATE requests.
	//'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
	//

> 	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the

Reviewer 3: does name always have a value though?  I thought names weren't generated until after admission.

Reviewer 6: Correct, if name is generated the name field will be unset during a CREATE because admission occurs before name generation. We should document this.  Note that for UPDATE, name is available.

Closely related. This whole chunk of documentation originated in CRD Validation Rules API documentation and we risk triplicating it or worse, so I've opened https://github.com/kubernetes/website/issues/39089 to consolidate the documentation.

> 	// request.object. No other metadata properties are accessible.

Reviewer 3: above it says `object`, here it says `request.object`.  Which is correct?

Author: My understanding is that `object` is just a convenience alias to `request.object`, but I changed this to `object` for consistency.

Reviewer 6: For ease of use in ValidationAdmissionPolicy, we _moved_ `object` out of request top level (`request.object` is not allowed). We did the same for `oldObject`. It should be trivial to use the same bit of code here so that `object` / `oldObject` work exactly the same in both APIs.

Reviewer 3: > For ease of use in ValidationAdmissionPolicy, we moved object from request to top level (request.object is not allowed).

This would work for me.  Allowing both seems strange to me.

Author: IMO it would be simpler to leave the `request` object alone, and make `object` an alias (or copy, if CEL doesn't have that notion). That way I can just look at the API documentation for `AdmissionReview` to know what fields are available.

Otherwise, the documentation will need to say something like "request is basically an AdmissionReview struct, with the following differences ..."

Reviewer 6: I honestly don't feel super strong about if object should be top level or not.  I'm less excited about having two ways to access the same variable (object and request.object). Maybe more pressing is that whatever we decide to do here is kept in sync with ValidationAdmissionPolicy (either by changing ValidationAdmissionPolicy or conforming to what it does).

> 	// request.object. No other metadata properties are accessible.

Reviewer 3: why is access to other metadata fields like labels, annotations, finalizers, ownerReferences restricted?

Reviewer 6: I don't think it should be. 

There's a back story here: This documentation chunk originated in CRD Validation Rules.  For CRD Validation Rules, we did not allow validation of metadata (except for name) to adhere to [rule #4 structural schemas](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema): "if metadata is specified, then only restrictions on metadata.name and metadata.generateName are allowed.".  This rule pre-dates CEL (it is also true for OpenAPIv3 validations) and ensures that metadata is uniform across resources (names being the exception which has it's own back story).

The rationale for CRD Validation Rules does not directly apply to admission rules which historically have been allowed to validate all metadata fields.

Reviewer 6: We need to fix this documentation on ValidatingAdmissionPolicy as well (the implementation does allow access to all metadata fields). I've made a note on https://github.com/kubernetes/enhancements/pull/3717/files#r1085742827 to get this fixed.

Author: Ack. I removed most of this comment block, and just put a reference to https://github.com/kubernetes/website/issues/39089 instead.

	//
	// Equality on arrays with list type of 'set' or 'map' ignores element order, i.e. [1, 2] == [2, 1].
	// Concatenation on arrays with x-kubernetes-list-type use the semantics of the list type:
	//   - 'set': `X + Y` performs a union where the array positions of all elements in `X` are preserved and
	//     non-intersecting elements in `Y` are appended, retaining their partial order.
	//   - 'map': `X + Y` performs a merge where the array positions of all keys in `X` are preserved but the values
	//     are overwritten by values in `Y` when the key sets of `X` and `Y` intersect. Elements in `Y` with
	//     non-intersecting keys are appended, retaining their partial order.
	// Required.
	Expression string `json:"expression"`
}
```

The match condition expression is evaluated by the same libraries as those used for CEL
ValidatingAdmissionPolicy. The only difference in expressions is the availability of the `params`
variable. Expressions requiring access to additional information outside the AdmissionRequest must
be performed in the webhook, and are out of scope for this proposal.

### Risks and Mitigations

> #### Security

Reviewer 3: it's glossed over here, but using expressions based on qualities of object that are not required for proper functioning make those admission checks optional.  A common example is the object label selection we already have.  It's worth enumerating the risk here and pointing out that the best mitigation is to couple your match condition to fields that are required for semantic efficacy of the resource and assume the user can lie about every non-functional aspect of the resource.

Author: Yes, I like that phrasing. I added a paragraph about this. PTAL.

**Risk: Attacker adds or changes a match condition to weaken an admission policy.**

This is does not represent a new threat, as doing so would require update access to the admission
registration object, and with that permission an attacker could already disable the policy through
manipulating match rules, namespace selector, or object selector (or reroute the webhook entirely).

**Risk: Logic error in match condition expression.**

Currently the match conditions must be encoded in the webhook backend itself. Moving the logic
into a CEL expression does not materially increase the risk of a logic bug.

#### Debugability

We do not normally log, audit, or emit an event when a webhook is out-of-scope for a request, and
the same will _mostly_ be true for match conditions.

At [log level V(5)](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use),
we will emit a log when a request that would otherwise be in-scope for a webhook is excluded for a
non-matching match condition.

> Short of increasing log verbosity, the recommended debug strategy is to capture or reproduce a

Reviewer 3: completely outside the scope of this KEP, but a thought for future-us.  Perhaps we could create a dry-run variant that includes information about the processing chain a particular request takes for each filter and admission plugin.  That could be neat.

Reviewer: Isn't this what tracing is for?

Reviewer 2: You could take the APF strategy and just put UIDs in the response header.

> pods with NFS volumes need to be checked.

Currently, there is no way to achieve this. Many webhook implementations today start by checking
that the request is within scope, and return early if it's not. This adds latency and an additional
failure point to irrelevant requests. This example requires an external integration, and thus is not
a candidate for migration to CEL `ValidatingAdmissionPolicy`.

With match conditions, the expressions can check whether the request object is in-scope for the
webhook:

```yaml
rules:
  - operations: ['CREATE']
    apiGroups: '' # core
    apiVersions: '*'
    resources: 'pods'
matchConditions:
  # Only include pods with an NFS volume.
  - name: 'nfs-volume-present'
    expression: 'object.spec.volumes.exists(v, v.has(nfs))'
```

### Goals

1. Provide a filtering mechanism for excluding requests from an admission webhook
2. Maintain consistency with `ValidatingAdmissionPolicy`

### Non-Goals

* Provide a mechanism to exclude requests from all webhooks.

## Proposal

### API

Both `ValidatingWebhook` and `MutatingWebhook` (in `admissionregistration.k8s.io`) will be updated
with a new `MatchConditions` field:

> ```go

Reviewer 2: nit: mixed tabs and spaces in this block.
